### PR TITLE
Adds: The C-Foam launcher

### PR DESCRIPTION
**Pull request by DGamerL**
## What Does This PR Do
Adds the GTFO inspired C-Foam launcher!
It shoots blobs of foam, which will:
- When hitting airlocks, it will create foam barriers that will block the airlock from being opened. You can shoot the airlock again to reinforce the blockage, up to 4 times. The last blob of foam will add 50 integrity, while the other 3 will add 25, for a total of 150 integrity.
- When hitting people, it will add a status effect which will slow them for the duration. For every blob of foam they get hit, they will be slowed for 10 seconds up to a maximum of 50 seconds. Your attack cooldown also gets slowed. If you are at the maximum foam level, you get stunned and knocked down for 4 seconds.

Can be refilled by printing cartridges from a hacked autolathe. Each cartridge contains 12 shots, costs 5000 metal and 2000 glass.
Obtainable from the uplink for 25 TC
It has 12 shots in the magazine

You cannot hotswap magazines.

It is C class.

## Why It's Good For The Game
It's a nonlethal tool to stop chases. It has some combat capabilities but it still costs 25 TC, so you become quite limited in your other options.

## Images of changes

image

## Testing
See above

## Changelog
:cl:
add: Adds the C-Foam launcher to the uplink for 25 TC!
add: Airlocks hit will be blocked up and required to have the foam blockage be broken
add: Humanoids (including aliens) will be slowed down, every hit will slow for 10 seconds
/:cl:

### CODE REVIEW

**Comment by Inferno1882** [comment]
I saw this PR ingame and went "I need some C-Foam on this door"
Should this be an R&D item as well, obviously behind an Illegals level?

**Comment by TravisAngeI** [comment]
> I saw this PR ingame and went "I need some C-Foam on this door" Should this be an R&D item as well, obviously behind an Illegals level?

probably a weaker version of it where the foam takes less hits to down, i could see officers using it to effectively have a pocket bolty-AI but that might just be my own paranoia

**Comment by Fordoxia** [comment]
> > I saw this PR ingame and went "I need some C-Foam on this door" Should this be an R&D item as well, obviously behind an Illegals level?
> 
> probably a weaker version of it where the foam takes less hits to down, i could see officers using it to effectively have a pocket bolty-AI but that might just be my own paranoia

They can already effectively do this with an experimental welding tool (or a regular one if they want to carry welding protection or get the eyes.

**Comment by Daylight2** [comment]
Looking at the images, I don't think any of these communicate 'I have foam on me' at all. Unless you're going for a high tech net aesthetic, instead of foam?

**Comment by DGamerL** [comment]
> I saw this PR ingame and went "I need some C-Foam on this door" Should this be an R&D item as well, obviously behind an Illegals level?

It's something that I am thinking about!



> Looking at the images, I don't think any of these communicate 'I have foam on me' at all. Unless you're going for a high tech net aesthetic, instead of foam?

image

File: code/__DEFINES/is_helpers.dm [hunk 1]
```diff
@@ -50,6 +50,8 @@
 
 #define ismachinery(A) (istype(A, /obj/machinery))
 
+#define isairlock(A) (istype(A, /obj/machinery/door))
+
```
Comment (Sirryan2002) [inline code/__DEFINES/is_helpers.dm:54]: I mean technically, mineral doors aren't airlocks...

File: code/game/machinery/doors/door.dm [hunk 2]
```diff
@@ -571,3 +581,23 @@
 
 	for(var/obj/airlock_filler_object/filler as anything in fillers)
 		filler.set_opacity(opacity)
+
+/obj/machinery/door/proc/foam_up(amount)
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:585]: document this proc please!

File: code/game/machinery/doors/door.dm [hunk 3]
```diff
@@ -571,3 +581,23 @@
...
+	if(!foam_level)
+		new /obj/structure/barricade/foam(get_turf(src))
+		foam_level++
+		return
+
+	if(foam_level == 5)
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:591]: convert `5` into a define please

File: code/modules/mob/living/carbon/carbon_defense.dm [hunk 4]
```diff
@@ -76,3 +76,6 @@
 	if(!affecting) //bruh where's your chest
 		return FALSE
 	apply_damage(damage, BRUTE, affecting)
+
+/mob/living/carbon/proc/foam_up(amount)
```
Comment (Sirryan2002) [inline code/modules/mob/living/carbon/carbon_defense.dm:80]: document this proc please

**Review by Sirryan2002 — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/door.dm (hunk 3 above)
Comment (DGamerL) [reply code/game/machinery/doors/door.dm:591]: Done!

File: code/game/machinery/doors/door.dm (hunk 2 above)
Comment (DGamerL) [reply code/game/machinery/doors/door.dm:585]: Done!

File: code/__DEFINES/is_helpers.dm (hunk 1 above)
Comment (DGamerL) [reply code/__DEFINES/is_helpers.dm:54]: Oh god mineral airlocks exist, I'll have to fix that

File: code/modules/mob/living/carbon/carbon_defense.dm (hunk 4 above)
Comment (DGamerL) [reply code/modules/mob/living/carbon/carbon_defense.dm:80]: Done!

**Comment by Scribble-Sheep** [comment]
I imagine this is meant to look like that expanding spray foam insulation stuff?

**Review by Sirryan2002 — APPROVED** [review]

**Review by Burzah — APPROVED** [review]
Looks good, just needs the merge conflict cleaned up.

**Review by DGamerL — COMMENTED** [review]
I sadly can't request changes to block this but this needs 2 more things done before merge:

- [x] Mineral doors
- [x] New sprites, either by mcramon or codersprites

I'll keep the conflict as an extra safety measure

**Comment by DGamerL** [comment]
I forgot draft PRs existed

**Comment by Sirryan2002** [comment]
@DGamerL updates on this PR?

**Review by S34NW — APPROVED** [review]

**Pull request merged**